stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,234 @@
+
+  // Variables required for all three Jacobian-adjoint products.
+  double norm_norm = alpha_log_norms(n_transitions);
+  double unnormed_marginal = alphas.col(n_transitions).sum();
+
+  std::vector<Eigen::VectorXd> kappa(n_transitions);
+  Eigen::VectorXd kappa_log_norms(n_transitions);
+  std::vector<double> grad_corr(n_transitions);
+
+  if (n_transitions > 0) {
+    kappa[n_transitions - 1] = Eigen::VectorXd::Ones(n_states);
+    kappa_log_norms(n_transitions - 1) = 0;
+    grad_corr[n_transitions - 1]
+        = std::exp(alpha_log_norms(n_transitions - 1) - norm_norm);
+  }
+
+  for (int n = n_transitions - 2; n >= 0; --n) {
+    kappa[n] = Gamma_dbl * (omegas.col(n + 2).cwiseProduct(kappa[n + 1]));
+
+    double norm = kappa[n].maxCoeff();
+    kappa[n] /= norm;
+    kappa_log_norms(n) = std::log(norm) + kappa_log_norms(n + 1);
+    grad_corr[n]
+        = std::exp(alpha_log_norms(n) + kappa_log_norms(n) - norm_norm);
+  }
+
+  if (!is_constant_all<T_Gamma>::value) {
+    Eigen::MatrixXd Gamma_jacad(n_states, n_states);
+    Gamma_jacad.setZero();
+
+    if (n_transitions != 0) {
+      for (int n = n_transitions - 1; n >= 0; --n) {
+        Gamma_jacad += (grad_corr[n] * kappa[n].cwiseProduct(omegas.col(n + 1))
+                        * alphas.col(n).transpose())
+                           .transpose();
+      }
+    }
+
+    Gamma_jacad /= unnormed_marginal;
+    ops_partials.edge2_.partials_ = Gamma_jacad;
+  }
+
+  bool sensitivities_for_omega_or_rho
+      = (!is_constant_all<T_omega>::value) || (!is_constant_all<T_rho>::value);
+
+  // boundary terms
+  if (sensitivities_for_omega_or_rho) {
+    Eigen::MatrixXd log_omega_jacad(n_states, n_transitions + 1);
+    log_omega_jacad.setZero();
+
+    if (!is_constant_all<T_omega>::value) {
+      for (int n = n_transitions - 1; n >= 0; --n)
+        log_omega_jacad.col(n + 1)
+            = grad_corr[n]
+              * kappa[n].cwiseProduct(Gamma_dbl.transpose() * alphas.col(n));
+    }
+
+    // Boundary terms
+    // TO DO (charlesm93): find a better solution that the if loop
+    // for the case with 0 transitions.
+    double grad_corr_boundary;
+    Eigen::VectorXd c;
+    if (n_transitions != 0) {
+      grad_corr_boundary = std::exp(kappa_log_norms(0) - norm_norm);
+      c = Gamma_dbl * omegas.col(1).cwiseProduct(kappa[0]);
+    }
+
+    if (!is_constant_all<T_omega>::value) {
+      if (n_transitions != 0) {
+        log_omega_jacad.col(0)
+            = grad_corr_boundary * c.cwiseProduct(value_of_rec(rho));
+        log_omega_jacad
+            = log_omega_jacad.cwiseProduct(omegas / unnormed_marginal);
+      } else {
+        log_omega_jacad.col(0) = omegas.col(0).cwiseProduct(value_of_rec(rho))
+                                 / exp(value_of_rec(log_marginal_density));
+      }
+      ops_partials.edge1_.partials_ = log_omega_jacad;
+    }
+
+    if (!is_constant_all<T_rho>::value) {
+      if (n_transitions != 0) {
```

Comment:
If `n_transitions == 0` isn't just split out at the top, push branching as far down as possible so that the parallel assignment structure is clear.

```
ops_partials.edge3_.partials_
  = (n_transitions == 0)
  ? omegas.col(0) / exp(value_of_rec(log_marginal_density))
  : grad_corr_boundary * c.cwiseProduct(omegas.col(0)) / unnormed_marginal;
```

Reply:
Handling the `n_transitions == 0` at the top now.